packages/ckeditor5-utils/src/dom/rect.ts

```diff
@@ -261,16 +267,53 @@ export default class Rect {
 
```

Comment:
I think this code should at least `warn` if `!parent`. Using `getVisible()` with an abstract rect that has no clear `_source` such as DOM range or DOM element makes little sense. It's always fully visible. But doing so in the code may produce confusing results that are **very hard to debug**.

FYI: I tried to throw here to see how many tests will fail and there are 30+ in utils+ui. They are mostly caused by places like this one where we mock stuff instead of using some real elements

https://github.com/ckeditor/ckeditor5/blob/d8672c397807971111f5350aa020853159c41082/packages/ckeditor5-ui/tests/panel/balloon/balloonpanelview.js#L1351-L1358

Anyway, this is not a blocker but IMO makes for a serious follow-up.

Reply:
Agree, noticed it also... well it's caused cause `getVisible` will now works only real DOM elements, and finding intersection between two `Rect`'s should only be done using `getIntersection`.

Reply:
Follow-up :point\_right: #14964